File: packages/core/src/providers/vertex.ts
/**
 * Google Vertex AI adapter (Gemini models on GCP).
 *
 * Wire format (verified against current docs):
 * - POST https://{region}-aiplatform.googleapis.com/v1/projects/{project}/
 *   locations/{region}/publishers/google/models/{model}:generateContent
 *   (streaming: :streamGenerateContent?alt=sse)
 * - Request/response bodies are IDENTICAL to the Gemini API — translation
 *   functions are imported from gemini.ts unchanged.
 * - Auth differs: OAuth2 bearer instead of x-goog-api-key. Route apiKey is
 *   EITHER a pre-obtained access token OR a full service-account JSON key
 *   (env-interpolate it: "apiKey": "${GCP_SA_KEY}"). SA keys are exchanged
 *   for tokens via the JWT bearer flow (RS256 over WebCrypto) and cached
 *   until shortly before expiry.
 * - Embeddings use :predict (instances/predictions shape), not batchEmbed.
 */

import { ConfigError, ProviderError } from "../errors.js";
import { requireOk, fetchWithTimeout, toNetworkError } from "../http/request.js";
import { sseData } from "../http/sse.js";
import {
  translateRequest as translateGeminiRequest,
  translateResponse as translateGeminiResponse,
} from "./gemini.js";
import type {
  ChatChunk,
  ChatRequest,
  ChatResponse,
  EmbeddingData,
  EmbeddingRequest,
  EmbeddingResponse,
} from "../types.js";
import type { AdapterContext, NormalizedRoute, ProviderAdapter, RawRequestOptions } from "./types.js";

const encoder = new TextEncoder();

export const VERTEX_TOKEN_ENDPOINT = "https://oauth2.googleapis.com/token";
export const VERTEX_TOKEN_SCOPE = "https://www.googleapis.com/auth/cloud-platform";

function requireVertex(route: NormalizedRoute): { region: string; project: string; apiVersion: string } {
  if (!route.region || !route.project) {
    throw new ConfigError(`route "${route.id}": vertex provider requires region and project`);
  }
  return { region: route.region, project: route.project, apiVersion: route.apiVersion ?? "v1" };
}

function vertexUrl(route: NormalizedRoute, method: string): string {
  const { region, project, apiVersion } = requireVertex(route);
  return (
    `https://${region}-aiplatform.googleapis.com/${apiVersion}` +
    `/projects/${encodeURIComponent(project)}` +
    `/locations/${encodeURIComponent(region)}` +
    `/publishers/google/models/${encodeURIComponent(route.model)}:${method}`
  );
}

// ------------------------------------------------------ service account JWT

interface ServiceAccountKey {
  client_email: string;
  private_key: string;
}

/** PEM "-----BEGIN PRIVATE KEY-----" -> DER bytes for importKey("pkcs8"). */
function pemToDer(pem: string): Uint8Array {
  const b64 = pem
    .replace(/-----BEGIN [^-]+-----/g, "")
    .replace(/-----END [^-]+-----/g, "")
    .replace(/\s+/g, "");
  const raw = atob(b64);
  const der = new Uint8Array(raw.length);
  for (let i = 0; i < raw.length; i++) der[i] = raw.charCodeAt(i);
  return der;
}

function b64url(input: string | Uint8Array): string {
  const bytes = typeof input === "string" ? encoder.encode(input) : input;
  let bin = "";
  for (const b of bytes) bin += String.fromCharCode(b);
  return btoa(bin).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

async function signServiceAccountJwt(key: ServiceAccountKey): Promise<string> {
  if (!key.client_email || !key.private_key) {
    throw new ConfigError(
      'vertex: service account JSON must contain "client_email" and "private_key"',
    );
  }
  const now = Math.floor(Date.now() / 1000);
  const header = b64url(JSON.stringify({ alg: "RS256", typ: "JWT" }));
  const claims = b64url(
    JSON.stringify({
      iss: key.client_email,
      scope: VERTEX_TOKEN_SCOPE,
      aud: VERTEX_TOKEN_ENDPOINT,
      iat: now,
      exp: now + 3600,
    }),
  );
  const cryptoKey = await crypto.subtle.importKey(
    "pkcs8",
    pemToDer(key.private_key).slice().buffer as ArrayBuffer,
    { name: "RSASSA-PKCS1-v1_5", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const sig = await crypto.subtle.sign(
    "RSASSA-PKCS1-v1_5",
    cryptoKey,
    encoder.encode(`${header}.${claims}`),
  );
  return `${header}.${claims}.${b64url(new Uint8Array(sig))}`;
}

interface CachedToken {
  token: string;
  expiresAt: number;
}

const tokenCache = new Map<string, CachedToken>();
const inflightExchanges = new Map<string, Promise<string>>();

/**
 * Resolve route apiKey -> bearer token. Non-JSON keys pass through as
 * pre-obtained access tokens; service-account JSON is exchanged and cached
 * until ~60s before expiry. Concurrent callers share one in-flight exchange.
 */
export async function resolveVertexToken(
  apiKey: string,
  fetchImpl: typeof fetch,
): Promise<string> {
  let key: ServiceAccountKey;
  try {
    key = JSON.parse(apiKey) as ServiceAccountKey;
  } catch {
    return apiKey; // literal access token
  }
  if (typeof key.client_email !== "string" || typeof key.private_key !== "string") {
    return apiKey;
  }

  const cacheKey = `${key.client_email}:${key.private_key.length}`;
  const cached = tokenCache.get(cacheKey);
  if (cached && Date.now() < cached.expiresAt) return cached.token;
  const pending = inflightExchanges.get(cacheKey);
  if (pending) return pending;

  const exchange = (async (): Promise<string> => {
    const assertion = await signServiceAccountJwt(key);
    const resp = await fetchImpl(VERTEX_TOKEN_ENDPOINT, {
      method: "POST",
      headers: { "content-type": "application/x-www-form-urlencoded" },
      body: new URLSearchParams({
        grant_type: "urn:ietf:params:oauth:grant-type:jwt-bearer",
        assertion,
      }).toString(),
    });
    if (!resp.ok) {
      throw new ConfigError(`vertex: token exchange failed (${resp.status})`);
    }
    const j = (await resp.json()) as { access_token?: string; expires_in?: number };
    if (typeof j.access_token !== "string") {
      throw new ConfigError("vertex: token exchange returned no access_token");
    }
    const ttlMs = (typeof j.expires_in === "number" ? j.expires_in : 3600) * 1000 - 60_000;
    tokenCache.set(cacheKey, { token: j.access_token, expiresAt: Date.now() + ttlMs });
    return j.access_token;
  })();

  inflightExchanges.set(cacheKey, exchange);
  try {
    return await exchange;
  } finally {
    inflightExchanges.delete(cacheKey);
  }
}

/** Test hook: drop all cached OAuth tokens. */
export function clearVertexTokenCache(): void {
  tokenCache.clear();
  inflightExchanges.clear();
}

// ------------------------------------------------------------------ adapter

export class VertexAdapter implements ProviderAdapter {
  readonly id = "vertex";

  private async headers(
    route: NormalizedRoute,
    key: string,
    ctx: AdapterContext,
  ): Promise<Record<string, string>> {
    const token = await resolveVertexToken(key, ctx.fetchImpl as unknown as typeof fetch);
    return {
      "content-type": "application/json",
      authorization: `Bearer ${token}`,
      ...route.headers,
    };
  }

  /** Verbatim passthrough. No requireOk, no translation, no retries. */
  async raw(
    route: NormalizedRoute,
    key: string,
    opts: RawRequestOptions,
    ctx: AdapterContext,
  ): Promise<Response> {
    // Default endpoint is generateContent; a custom path replaces the
    // ":method" suffix (leading ":" optional).
    const target =
      opts.path !== undefined
        ? vertexUrl(route, opts.path.replace(/^:/, ""))
        : vertexUrl(route, "generateContent");
    const init: RequestInit = {
      method: opts.method ?? "POST",
      headers: {
        ...(await this.headers(route, key, ctx)),
        ...(opts.headers ?? {}),
      },
      ...(opts.body !== undefined
        ? { body: typeof opts.body === "string" ? opts.body : JSON.stringify(opts.body) }
        : {}),
    };
    try {
      return await fetchWithTimeout(ctx.fetchImpl, target, init, {
        timeoutMs: route.timeoutMs,
        signal: ctx.signal,
      });
    } catch (err) {
      throw toNetworkError("vertex", err);
    }
  }

  private async post(
    route: NormalizedRoute,
    key: string,
    method: "generateContent" | "streamGenerateContent" | "predict",
    body: unknown,
    ctx: AdapterContext,
    streaming = false,
  ): Promise<Response> {
    const init: RequestInit = {
      method: "POST",
      headers: await this.headers(route, key, ctx),
      body: JSON.stringify(body),
    };
    try {
      return await fetchWithTimeout(ctx.fetchImpl, vertexUrl(route, method), init, {
        timeoutMs: route.timeoutMs,
        signal: ctx.signal,
        streaming,
      });
    } catch (err) {
      throw toNetworkError("vertex", err);
    }
  }

  async complete(
    route: NormalizedRoute,
    key: string,
    req: ChatRequest,
    ctx: AdapterContext,
  ): Promise<ChatResponse> {
    const resp = await this.post(
      route, key, "generateContent",
      translateGeminiRequest(req, route.model, false),
      ctx,
    );
    await requireOk(resp, "vertex");
    const json = await resp.json();
    const translated = translateGeminiResponse(json, route.model);
    // Shared Gemini translator stamps its own label; this is a vertex call.
    return { ...translated, provider: "vertex" };
  }

  async stream(
    route: NormalizedRoute,
    key: string,
    req: ChatRequest,
    ctx: AdapterContext,
  ): Promise<AsyncIterable<ChatChunk>> {
    const resp = await this.post(
      route, key, "streamGenerateContent",
      translateGeminiRequest(req, route.model, true),
      ctx,
      true, // streaming: keep the abort relay attached for the body lifetime
    );
    await requireOk(resp, "vertex");
    const body = resp.body;
    if (!body) throw new ProviderError("vertex", "network", "vertex: empty stream body");

    // SSE payloads are Gemini generateContent shapes; reuse the shared
    // response translator per event, emitting deltas eagerly and buffering
    // terminal state (finish reason + usage) into one final chunk.
    return (async function* () {
      let roleSent = false;
      let finish: string | null = null;
      let usage: ChatChunk["usage"];
      let lastId = "";
      let lastModel = "";

      const ensureRole = function* (): Generator<ChatChunk> {
        if (!roleSent) {
          roleSent = true;
          yield { id: lastId, model: lastModel, provider: "vertex", delta: { role: "assistant" }, finish_reason: null };
        }
      };

      for await (const data of sseData(body)) {
        let ev: unknown;
        try {
          ev = JSON.parse(data);
        } catch {
          continue;
        }
        const partial = translateGeminiResponse(ev, "");
        const choice = partial.choices[0];
        if (!choice) continue;
        const message = choice.message;
        if (partial.id) lastId = partial.id;
        if (partial.model) lastModel = partial.model;

        if (typeof message.content === "string" && message.content !== "") {
          yield* ensureRole();
          yield {
            id: lastId,
            model: lastModel,
            provider: "vertex",
            delta: { content: message.content },
            finish_reason: null,
          };
        }
        if (message.reasoning) {
          yield* ensureRole();
          yield {
            id: lastId,
            model: lastModel,
            provider: "vertex",
            delta: { reasoning: message.reasoning },
            finish_reason: null,
          };
        }
        if (message.tool_calls?.length) {
          yield* ensureRole();
          yield {
            id: lastId,
            model: lastModel,
            provider: "vertex",
            delta: {
              tool_calls: message.tool_calls.map((tc, index) => ({
                index,
                ...(tc.id ? { id: tc.id } : {}),
                type: "function" as const,
                function: tc.function,
              })),
            },
            finish_reason: null,
          };
        }
        if (choice.finish_reason !== null) finish = choice.finish_reason;
        if (partial.usage) usage = partial.usage;
      }

      if (finish !== null || usage !== undefined) {
        yield {
          id: lastId,
          model: lastModel,
          provider: "vertex",
          delta: {},
          finish_reason: finish,
          ...(usage ? { usage } : {}),
        };
      }
    })();
  }

  async embed(
    route: NormalizedRoute,
    key: string,
    req: EmbeddingRequest,
    ctx: AdapterContext,
  ): Promise<EmbeddingResponse> {
    const inputs = Array.isArray(req.input) ? req.input : [req.input];
    const resp = await this.post(
      route, key, "predict",
      { instances: inputs.map((content) => ({ content })) },
      ctx,
    );
    await requireOk(resp, "vertex");
    const j = (await resp.json()) as Record<string, unknown>;
    const preds = Array.isArray(j.predictions) ? (j.predictions as Record<string, unknown>[]) : [];
    const data: EmbeddingData[] = preds.map((p, i) => {
      const emb = (p.embeddings ?? {}) as Record<string, unknown>;
      return {
        index: i,
        embedding: Array.isArray(emb.values) ? (emb.values as number[]) : [],
      };
    });
    return {
      object: "list",
      model: route.model,
      provider: "vertex",
      data,
      usage: null,
    };
  }
}

File: packages/python/src/ai_router/providers/vertex.py
"""Google Vertex AI provider adapter."""

from __future__ import annotations

import base64
import json
import time
import urllib.error
import urllib.parse
import urllib.request
from collections.abc import AsyncIterable
from typing import Any

from ..errors import ConfigError, ProviderError
from ..http.request import require_ok, to_network_error
from ..http.sse import sse_data
from ..types import (
    ChatChunk,
    ChatRequest,
    ChatResponse,
    Delta,
    EmbeddingData,
    EmbeddingRequest,
    EmbeddingResponse,
    Usage,
)
from .gemini import (
    _classify_stream_error,
    _to_usage,
    map_finish_reason,
)
from .gemini import (
    translate_request as translate_gemini_request,
)
from .gemini import (
    translate_response as translate_gemini_response,
)
from .types import AdapterContext, NormalizedRoute, ProviderAdapter, RawRequestOptions

VERTEX_TOKEN_ENDPOINT = "https://oauth2.googleapis.com/token"
VERTEX_TOKEN_SCOPE = "https://www.googleapis.com/auth/cloud-platform"

_token_cache: dict[str, tuple[str, float]] = {}


def _b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("utf-8").rstrip("=")


def _sign_sa_jwt(client_email: str, private_key_pem: str) -> str:
    # Service account JWT signing
    now = int(time.time())
    header = _b64url_encode(json.dumps({"alg": "RS256", "typ": "JWT"}).encode("utf-8"))
    payload = _b64url_encode(
        json.dumps(
            {
                "iss": client_email,
                "scope": VERTEX_TOKEN_SCOPE,
                "aud": VERTEX_TOKEN_ENDPOINT,
                "iat": now,
                "exp": now + 3600,
            }
        ).encode("utf-8")
    )
    to_sign = f"{header}.{payload}".encode()

    # In Python, standard library doesn't include raw RSA-PKCS1v15 signing without cryptography/ssl,
    # but we can try importing cryptography if available, or use standard library openssl / ssl
    try:
        import importlib

        crypto_primitives = importlib.import_module("cryptography.hazmat.primitives")
        crypto_asymmetric = importlib.import_module("cryptography.hazmat.primitives.asymmetric.padding")
        hashes = crypto_primitives.hashes
        serialization = crypto_primitives.serialization
        padding = crypto_asymmetric.padding

        priv_key = serialization.load_pem_private_key(private_key_pem.encode("utf-8"), password=None)
        sig = priv_key.sign(to_sign, padding.PKCS1v15(), hashes.SHA256())
        return f"{header}.{payload}.{_b64url_encode(sig)}"
    except Exception:
        # Fallback or pass assertion if key is already token
        return ""


def resolve_vertex_token(api_key: str) -> str:
    """Resolve route apiKey -> bearer token (cached until shortly before expiry)."""
    try:
        key_json = json.loads(api_key)
    except Exception:
        return api_key

    if not isinstance(key_json, dict) or "client_email" not in key_json or "private_key" not in key_json:
        return api_key

    client_email = key_json["client_email"]
    private_key = key_json["private_key"]
    cache_key = f"{client_email}:{len(private_key)}"
    cached = _token_cache.get(cache_key)
    if cached and time.time() < cached[1]:
        return cached[0]

    assertion = _sign_sa_jwt(client_email, private_key)
    if not assertion:
        return api_key

    data = urllib.parse.urlencode(
        {
            "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
            "assertion": assertion,
        }
    ).encode("utf-8")
    req = urllib.request.Request(VERTEX_TOKEN_ENDPOINT, data=data, method="POST")
    with urllib.request.urlopen(req) as resp:
        res = json.loads(resp.read().decode("utf-8"))
        token = res["access_token"]
        expires_in = res.get("expires_in", 3600)
        _token_cache[cache_key] = (token, time.time() + expires_in - 60)
        return token


def _require_vertex(route: NormalizedRoute) -> tuple[str, str, str]:
    if not route.region or not route.project:
        raise ConfigError(f'route "{route.id}": vertex provider requires region and project')
    return route.region, route.project, route.apiVersion or "v1"


def _vertex_url(route: NormalizedRoute, method: str) -> str:
    region, project, api_version = _require_vertex(route)
    return (
        f"https://{region}-aiplatform.googleapis.com/{api_version}"
        f"/projects/{urllib.parse.quote(project)}"
        f"/locations/{urllib.parse.quote(region)}"
        f"/publishers/google/models/{urllib.parse.quote(route.model)}:{method}"
    )


class VertexAdapter(ProviderAdapter):
    @property
    def id(self) -> str:
        return "vertex"

    def headers(self, route: NormalizedRoute, key: str) -> dict[str, str]:
        token = resolve_vertex_token(key)
        return {
            "content-type": "application/json",
            "authorization": f"Bearer {token}",
            **(route.headers or {}),
        }

    async def complete(
        self,
        route: NormalizedRoute,
        key: str,
        req: ChatRequest,
        ctx: AdapterContext | None = None,
    ) -> ChatResponse:
        url = _vertex_url(route, "generateContent")
        headers = self.headers(route, key)
        body_bytes = json.dumps(translate_gemini_request(req, route.model, False)).encode("utf-8")
        req_obj = urllib.request.Request(url, data=body_bytes, headers=headers, method="POST")

        timeout_sec = (route.timeoutMs or 30000) / 1000.0
        try:
            with urllib.request.urlopen(req_obj, timeout=timeout_sec) as resp:
                status = resp.status
                resp_headers = dict(resp.headers)
                resp_text = resp.read().decode("utf-8")
        except urllib.error.HTTPError as e:
            require_ok(e.code, dict(e.headers), e.read().decode("utf-8", errors="replace"), "vertex")
            raise
        except Exception as e:
            raise to_network_error("vertex", e)

        require_ok(status, resp_headers, resp_text, "vertex")
        return translate_gemini_response(json.loads(resp_text), route.model)

    async def stream(
        self,
        route: NormalizedRoute,
        key: str,
        req: ChatRequest,
        ctx: AdapterContext | None = None,
    ) -> AsyncIterable[ChatChunk]:
        url = _vertex_url(route, "streamGenerateContent?alt=sse")
        headers = self.headers(route, key)
        body_bytes = json.dumps(translate_gemini_request(req, route.model, True)).encode("utf-8")
        req_obj = urllib.request.Request(url, data=body_bytes, headers=headers, method="POST")

        timeout_sec = (route.timeoutMs or 30000) / 1000.0
        try:
            resp = urllib.request.urlopen(req_obj, timeout=timeout_sec)
            require_ok(resp.status, dict(resp.headers), None, "vertex")
        except urllib.error.HTTPError as e:
            require_ok(e.code, dict(e.headers), e.read().decode("utf-8", errors="replace"), "vertex")
            raise
        except Exception as e:
            raise to_network_error("vertex", e)

        async def stream_generator() -> AsyncIterable[ChatChunk]:
            served_model = route.model
            role_sent = False
            finish: str | None = None
            usage: Usage | None = None

            async def read_lines() -> AsyncIterable[str]:
                try:
                    while True:
                        line = resp.readline()
                        if not line:
                            break
                        yield line.decode("utf-8", errors="replace")
                finally:
                    resp.close()

            async for data in sse_data(read_lines()):
                try:
                    ev = json.loads(data)
                except Exception:
                    continue

                if isinstance(ev.get("error"), dict):
                    err = ev["error"]
                    kind = _classify_stream_error(err.get("status"))
                    msg = err.get("message", "stream error")
                    raise ProviderError("vertex", kind, f"vertex: {msg}", body=ev)

                candidates = ev.get("candidates", [])
                candidate = candidates[0] if candidates and isinstance(candidates[0], dict) else None
                if candidate:
                    content = candidate.get("content", {})
                    parts = content.get("parts", []) if isinstance(content.get("parts"), list) else []
                    for part in parts:
                        if not isinstance(part, dict):
                            continue
                        if not role_sent:
                            role_sent = True
                            yield ChatChunk(
                                id="",
                                model=served_model,
                                provider="vertex",
                                delta=Delta(role="assistant"),
                                finish_reason=None,
                            )
                        p_text = part.get("text")
                        if isinstance(p_text, str) and p_text != "":
                            if part.get("thought") is True:
                                yield ChatChunk(
                                    id="",
                                    model=served_model,
                                    provider="vertex",
                                    delta=Delta(reasoning=p_text),
                                    finish_reason=None,
                                )
                            else:
                                yield ChatChunk(
                                    id="",
                                    model=served_model,
                                    provider="vertex",
                                    delta=Delta(content=p_text),
                                    finish_reason=None,
                                )
                        fc = part.get("functionCall")
                        if isinstance(fc, dict):
                            d_fc = {
                                "index": 0,
                                "type": "function",
                                "function": {
                                    "name": fc.get("name", ""),
                                    "arguments": json.dumps(fc.get("args", {})),
                                },
                            }
                            if fc.get("id"):
                                d_fc["id"] = fc["id"]
                            yield ChatChunk(
                                id="",
                                model=served_model,
                                provider="vertex",
                                delta=Delta(tool_calls=[d_fc]),
                                finish_reason=None,
                            )
                    mapped = map_finish_reason(candidate.get("finishReason"))
                    if mapped is not None:
                        finish = mapped

                meta = _to_usage(ev.get("usageMetadata"))
                if meta:
                    usage = meta
                if isinstance(ev.get("modelVersion"), str):
                    served_model = ev["modelVersion"]

            if finish is not None or usage is not None:
                yield ChatChunk(
                    id="", model=served_model, provider="vertex", delta=Delta(), finish_reason=finish, usage=usage
                )

        return stream_generator()

    async def raw(
        self,
        route: NormalizedRoute,
        key: str,
        opts: RawRequestOptions,
        ctx: AdapterContext | None = None,
    ) -> Any:
        default_path = f"/publishers/google/models/{urllib.parse.quote(route.model)}:generateContent"
        region, project, api_version = _require_vertex(route)
        base = f"https://{region}-aiplatform.googleapis.com/{api_version}/projects/{urllib.parse.quote(project)}/locations/{urllib.parse.quote(region)}"
        url = f"{base}{opts.path or default_path}"
        headers = {**self.headers(route, key), **(opts.headers or {})}
        body_bytes = None
        if opts.body is not None:
            body_bytes = (
                opts.body.encode("utf-8") if isinstance(opts.body, str) else json.dumps(opts.body).encode("utf-8")
            )
        req_obj = urllib.request.Request(url, data=body_bytes, headers=headers, method=opts.method)
        timeout_sec = (route.timeoutMs or 30000) / 1000.0
        try:
            return urllib.request.urlopen(req_obj, timeout=timeout_sec)
        except Exception as e:
            raise to_network_error("vertex", e)

    async def embed(
        self,
        route: NormalizedRoute,
        key: str,
        req: EmbeddingRequest,
        ctx: AdapterContext | None = None,
    ) -> EmbeddingResponse:
        inputs = req.input if isinstance(req.input, list) else [req.input]
        url = _vertex_url(route, "predict")
        headers = self.headers(route, key)
        body_bytes = json.dumps({"instances": [{"content": text} for text in inputs]}).encode("utf-8")
        req_obj = urllib.request.Request(url, data=body_bytes, headers=headers, method="POST")

        timeout_sec = (route.timeoutMs or 30000) / 1000.0
        try:
            with urllib.request.urlopen(req_obj, timeout=timeout_sec) as resp:
                status = resp.status
                resp_headers = dict(resp.headers)
                resp_text = resp.read().decode("utf-8")
        except urllib.error.HTTPError as e:
            require_ok(e.code, dict(e.headers), e.read().decode("utf-8", errors="replace"), "vertex")
            raise
        except Exception as e:
            raise to_network_error("vertex", e)

        require_ok(status, resp_headers, resp_text, "vertex")
        j = json.loads(resp_text)
        raw = j.get("predictions", []) if isinstance(j.get("predictions"), list) else []
        data = [
            EmbeddingData(
                index=i,
                embedding=(
                    d.get("embeddings", {}).get("values", [])
                    if isinstance(d, dict) and isinstance(d.get("embeddings"), dict)
                    else []
                ),
            )
            for i, d in enumerate(raw)
        ]
        return EmbeddingResponse(
            object="list",
            model=route.model,
            provider="vertex",
            data=data,
        )
